### ISSUE TITLE
Several blocks with the anchor are broken in the update

### ISSUE THREAD

**Issue by t-hamano** [WordPress/gutenberg#48232]
In WordPress 6.1.1, if the content contains blocks with the anchor, updating to WordPress 6.2 Beta2 will break some blocks.

I created the following content in WordPress 6.1.1:

```html
<!-- wp:paragraph -->
<p>Paragraph without anchor</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p id="anchor">Paragraph with anchor</p>
<!-- /wp:paragraph -->

<!-- wp:heading -->
<h2>Heading without anchor</h2>
<!-- /wp:heading -->

<!-- wp:heading -->
<h2 id="anchor">Heading with anchor</h2>
<!-- /wp:heading -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div id="anchor" class="wp-block-group"></div>
<!-- /wp:group -->
```

![6 1](https://user-images.githubusercontent.com/54422211/219936659-655f1365-3866-4b4f-bc10-0f4c78d3748c.png)

If you update this environment to WordPress 6.2, the markup will change as follows:

```html
<!-- wp:paragraph -->
<p>Paragraph without anchor</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p><p id="anchor">Paragraph with anchor</p></p>
<!-- /wp:paragraph -->

<!-- wp:heading -->
<h2 class="wp-block-heading">Heading without anchor</h2>
<!-- /wp:heading -->

<!-- wp:heading {"anchor":"anchor"} -->
<h2 class="wp-block-heading" id="anchor">Heading with anchor</h2>
<!-- /wp:heading -->

<!-- wp:group {"layout":{"type":"constrained"}} /-->
```

A comparison of the markups confirms the following three points:

- The paragraph block with the anchor are nested in the `p` element
- A comment about `anchor` attributes has been added to the heading block
- The anchor in the group block are missing

Now save the post, updating only the title, and reload the browser. Then the block will be broken as follows:

![6 2](https://user-images.githubusercontent.com/54422211/219936882-92bd21fd-520f-46c1-948a-fa67fedae902.png)

The browser console displays the following error:

```
react_devtools_backend.js:4012 Block validation: Block validation failed for `core/paragraph` ({name: 'core/paragraph', icon: {…}, keywords: Array(1), attributes: {…}, providesContext: {…}, …}).

Content generated by `save` function:

<p></p>

Content retrieved from post body:

<p><p id="anchor">Paragraph with anchor</p></p>

---------------

react_devtools_backend.js:4012 Block validation: Block validation failed for `core/group` ({name: 'core/group', icon: {…}, keywords: Array(4), attributes: {…}, providesContext: {…}, …}).

Content generated by `save` function:

<div class="wp-block-group"></div>

Content retrieved from post body:
```

**Comment by t-hamano** [comment]
This issue might be related to #44771.

- In the Gutenberg repository, start wp-env with WordPress 6.1.1.
- Disable the Gutenberg plugin.
- Insert paragraphs, headings, group blocks, etc. with the anchor into the post content.
- Enable the Gutenberg plugin.
- When you open the post in the code editor, the markup has changed.
- Comment out [this line](https://github.com/WordPress/gutenberg/blob/7f4bc290d92c661f8d0e7e259166bfdae702b043/lib/load.php#L155) added by #44771.
- Open the code editor and confirm that the markup has not changed.

As mentioned above, I think that the changes made in #44771 might have unintentionally affected static blocks.

**Comment by Mamaduka** [comment]
Cc @Soean

**Comment by Soean** [comment]
Maybe we should check, if the block has a render callback.

**Comment by t-hamano** [comment]
I think the render callback is not limited to dynamic blocks but is also used to rewrite output on the front end. For example, A _static_ block with a render callback will have extra comments added as follows:

```html
<!-- From -->

<!-- wp:heading -->
<h2 id="my-heading">Test</h2>
<!-- /wp:heading -->

<!-- wp:image {"sizeSlug":"large"} -->
<figure class="wp-block-image size-large" id="my-image"><img src="https://live.staticflickr.com/4572/38004374914_6b686d708e_b.jpg" alt=""/></figure>
<!-- /wp:image -->

<!-- To -->

<!-- wp:heading {"anchor":"my-heading"} -->
<h2 class="wp-block-heading" id="my-heading">Test</h2>
<!-- /wp:heading -->

<!-- wp:image {"sizeSlug":"large","anchor":"my-image"} -->
<figure class="wp-block-image size-large" id="my-image"><img src="https://live.staticflickr.com/4572/38004374914_6b686d708e_b.jpg" alt=""/></figure>
<!-- /wp:image -->
```

I think we should check if the `save` property of `registerBlockType()` is `null`, but the server side cannot determine that...

**Comment by t-hamano** [comment]
As far as I know, I think there is no way for PHP to determine if the block needs to save the anchor attribute as a comment. Therefore, I propose the following approach.

- Explicitly define the `anchor` attribute in the dynamic block's `block.json`
- Add the following description to the block.json schema and documentation
  > For the dynamic block that doesn't save any markup with the save function, it is necessary to define the anchor property with type string as attributes.

**Comment by t-hamano** [comment]
The dynamic block anchor support that caused this problem has been reverted in WordPress 6.2 (See [this comment](https://github.com/WordPress/gutenberg/pull/48438#issuecomment-1451887454)).

This problem needs to be fixed when re-adding anchor support for dynamic blocks in the future. Therefore, move to "Punted to 6.3" on the project board.

**Comment by t-hamano** [comment]
Update: This issue is being attempted to be fixed in #48438.

**Comment by ocean90** [comment]
I think the changes should have been reverted from the plugin too since this is about losing all the content in a group block with an anchor set. The block recovery won't help here either.
The missing content is also only visible once reloading the editor which is unexpected and thus might get unnoticed.

What are the next steps?

**Comment by t-hamano** [comment]
Update: I have left [a comment](https://github.com/WordPress/gutenberg/pull/48438#issuecomment-1573688395) on how this issue should be addressed in the PR that will correct this problem.

**Comment by t-hamano** [comment]
Update: As mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/48438#issuecomment-1573688395), I have submitted #51288 to remove anchor support for dynamic blocks in order to find a more ideal solution.

After this PR is merged, I would like to close this issue and submit a new issue to consider re-adding anchor support for dynamic blocks.

**Comment by t-hamano** [comment]
Merged #51288 which revertss anchor support for dynamic blocks. A new issue exploring how to re-add anchor support is #51402.
